rasa/utils/tensorflow/model_data_utils.py

```diff
@@ -24,12 +24,123 @@
+    )
+    dense_sequence_features, dense_sentence_features = message.get_dense_features(
+        attribute, featurizers
+    )
+
+    if dense_sequence_features is not None and sparse_sequence_features is not None:
+        if (
+            dense_sequence_features.features.shape[0]
+            != sparse_sequence_features.features.shape[0]
+        ):
+            raise ValueError(
+                f"Sequence dimensions for sparse and dense sequence features "
+                f"don't coincide in '{message.get(TEXT)}'"
+                f"for attribute '{attribute}'."
+            )
+    if dense_sentence_features is not None and sparse_sentence_features is not None:
+        if (
+            dense_sentence_features.features.shape[0]
+            != sparse_sentence_features.features.shape[0]
+        ):
+            raise ValueError(
+                f"Sequence dimensions for sparse and dense sentence features "
+                f"don't coincide in '{message.get(TEXT)}'"
+                f"for attribute '{attribute}'."
+            )
+
+    out = {}
+    if sparse_sentence_features is not None:
+        out[(True, SENTENCE)] = sparse_sentence_features.features
+    if sparse_sequence_features is not None:
+        out[(True, SEQUENCE)] = sparse_sequence_features.features
+    if dense_sentence_features is not None:
+        out[(False, SENTENCE)] = dense_sentence_features.features
+    if dense_sequence_features is not None:
+        out[(False, SEQUENCE)] = dense_sequence_features.features
+
+    return out
+
+
+def extract_attribute_features_from_all_messages(
+    messages: List[Message],
+    attribute: Text,
+    type: Optional[Text] = None,
+    featurizers: Optional[List[Text]] = None,
+) -> Tuple[List[FeatureArray], List[FeatureArray]]:
+    """Collects and combines the attribute related features from all messages.
+
+    Args:
+        messages: list of messages to extract features from
+        attribute: the only attribute which will be considered
+        type: if not None, the other type (sentence/sequence) will be ignored
+        featurizers: the featurizers to be considered
+    Returns:
+        Sequence level features and sentence level features. Each feature contains
+        FeatureArrays with sparse features first.
+    """
+    if (type is not None) and (type not in [SENTENCE, SEQUENCE]):
+        raise ValueError(f"Unknown type {type}")
+    # for each label_example, collect sparse and dense feature (matrices) in lists
+    collected_features: Dict[
+        Tuple[Text, Text], List[Union[np.ndarray, scipy.sparse.spmatrix]]
+    ] = dict()
+    for msg in messages:
+        sparse_type_to_feature = extract_attribute_features_from_message(
+            message=msg, attribute=attribute, featurizers=featurizers,
+        )
+        for (feat_is_sparse, feat_type), feat_mat in sparse_type_to_feature.items():
+            if feat_type not in [SEQUENCE, SENTENCE]:
+                raise ValueError(f"Unknown type {feat_type}")

```

Comment:
Will add the "expected ... found. ... " as described above. In the long term, it would be great if we could just have FeatureType / FeatureLevel Enums 😃